packages/common/src/converter/protobuf-payload-converters.ts

```diff
@@ -121,17 +121,57 @@ export class ProtobufJsonPayloadConverter extends ProtobufPayloadConverter {
       return undefined;
     }
 
-    const jsonValue = protoJsonSerializer.toProto3JSON(value);
-
-    return this.constructPayload({
-      messageTypeName: getNamespacedTypeName(value.$type),
-      message: encode(JSON.stringify(jsonValue)),
-    });
+    const hasBufferChanged = setBufferInGlobal();
+    try {
+      const jsonValue = protoJsonSerializer.toProto3JSON(value);
+
+      return this.constructPayload({
+        messageTypeName: getNamespacedTypeName(value.$type),
+        message: encode(JSON.stringify(jsonValue)),
+      });
+    } finally {
+      resetBufferInGlobal(hasBufferChanged);
+    }
   }
 
   public fromPayload<T>(content: Payload): T {
-    const { messageType, data } = this.validatePayload(content);
-    return protoJsonSerializer.fromProto3JSON(messageType, JSON.parse(decode(data))) as unknown as T;
+    const hasBufferChanged = setBufferInGlobal();
+    try {
+      const { messageType, data } = this.validatePayload(content);
+      const res = protoJsonSerializer.fromProto3JSON(messageType, JSON.parse(decode(data))) as unknown as T;
+      replaceBuffers(res);
+      return res;
+    } finally {
+      resetBufferInGlobal(hasBufferChanged);
+    }
+  }
+}
+
+function replaceBuffers<X>(obj: X) {
+  if (obj != null && typeof obj === 'object') {
+    for (const [key, value] of Object.entries(obj)) {
+      if (Buffer.isBuffer(value)) {
+        type T = keyof typeof obj;
+        // Need to copy. `Buffer` manages a pool slab, internally reused when Buffer objects are GC.
+        obj[key as T] = new Uint8Array(value) as any;
+      } else {
+        replaceBuffers(value);
+      }
+    }
+  }
+}
+
+function setBufferInGlobal(): boolean {
+  if (typeof globalThis.Buffer === 'undefined') {
```

Comment:
A. Thanks for fixing this bug of mine 😄
B. Did you consider adding Buffer to global in [overrideGlobals](https://github.com/temporalio/sdk-typescript/blob/0a792088e13930b9dd4f0fcbebf24fc0b6cc6b74/packages/workflow/src/worker-interface.ts#L24) and leave it there?

Reply:
We had a long discussion on whether to add Buffer to globals or not, and for now the decision is not to do it because it can make it more difficult to port workers to non-nodejs environments...

Reply:
I see, thanks!